src/JsonApi/Serializer/CollectionNormalizer.php

```diff
@@ -0,0 +1,169 @@
+use ApiPlatform\Core\Metadata\Property\Factory\PropertyMetadataFactoryInterface;
+use ApiPlatform\Core\Metadata\Resource\Factory\ResourceMetadataFactoryInterface;
+use ApiPlatform\Core\Serializer\ContextTrait;
+use ApiPlatform\Core\Util\IriHelper;
+use Symfony\Component\Serializer\Normalizer\NormalizerAwareInterface;
+use Symfony\Component\Serializer\Normalizer\NormalizerAwareTrait;
+use Symfony\Component\Serializer\Normalizer\NormalizerInterface;
+
+/**
+ * Normalizes collections in the JSON API format.
+ *
+ * @author Kevin Dunglas <dunglas@gmail.com>
+ * @author Hamza Amrouche <hamza@les-tilleuls.coop>
+ */
+final class CollectionNormalizer implements NormalizerInterface, NormalizerAwareInterface
+{
+    use ContextTrait;
+    use NormalizerAwareTrait;
+
+    const FORMAT = 'jsonapi';
+
+    private $resourceClassResolver;
+    private $pageParameterName;
+    private $resourceMetadataFactory;
+    private $propertyMetadataFactory;
+
+    public function __construct(ResourceClassResolverInterface $resourceClassResolver, ResourceMetadataFactoryInterface $resourceMetadataFactory, PropertyMetadataFactoryInterface $propertyMetadataFactory, string $pageParameterName)
+    {
+        $this->resourceClassResolver = $resourceClassResolver;
+        $this->pageParameterName = $pageParameterName;
+        $this->resourceMetadataFactory = $resourceMetadataFactory;
+        $this->propertyMetadataFactory = $propertyMetadataFactory;
+    }
+
+    /**
+     * {@inheritdoc}
+     */
+    public function supportsNormalization($data, $format = null)
+    {
+        return self::FORMAT === $format && (is_array($data) || ($data instanceof \Traversable));
+    }
+
+    /**
+     * {@inheritdoc}
+     */
+    public function normalize($data, $format = null, array $context = [])
+    {
+        $currentPage = $lastPage = $itemsPerPage = 1;
+
+        // If we are normalizing stuff one level down (i.e., an attribute which
+        // could be already an array)
+        $returnDataArray = [];
+        if (isset($context['api_sub_level'])) {
+            foreach ($data as $index => $obj) {
+                $returnDataArray['data'][][$index] = $this->normalizer->normalize($obj, $format, $context);
+            }
+
+            return $data;
+        }
+
+        $resourceClass = $this->resourceClassResolver->getResourceClass(
+            $data,
+            $context['resource_class'] ?? null,
+            true
+        );
+
+        $resourceMetadata = $this->resourceMetadataFactory->create($resourceClass);
+
+        $context = $this->initContext($resourceClass, $context);
+
+        $parsed = IriHelper::parseIri($context['uri'] ?? '/', $this->pageParameterName);
+        $paginated = $isPaginator = $data instanceof PaginatorInterface;
+
+        if ($isPaginator) {
+            $currentPage = $data->getCurrentPage();
+            $lastPage = $data->getLastPage();
+            $itemsPerPage = $data->getItemsPerPage();
+
+            $paginated = 1. !== $lastPage;
```

Comment:
https://github.com/api-platform/core/pull/1175/files#diff-5b154fa3efaf5776695baccd0a52daf4R66 initialized as integer? IMO pagination stuff should be integer-only because it doesn't make sense to have a float ^^ (maybe jsonapi specifies float though I haven't looked at it).